dartagnan/src/main/java/com/dat3m/dartagnan/program/memory/Address.java

```diff
@@ -1,33 +1,102 @@
+
+    private final HashMap<Integer,IConst> initialValues = new HashMap<>();
 
-    Address(int index){
-        super(BigInteger.valueOf(index), -1);
+    Address(int index, int s) {
         this.index = index;
+        size = s;
+    }
+
+    /**
+     * @return
+     * Number of fields in this array.
+     */
+    public int size() {
+        return size;
+    }
+
+    /**
+     * Updates the initial value at a certain field of this array.
+     * @param offset
+     * Non-negative number of fields before the target.
+     * @return
+     * Readable value at the start of each execution.
+     */
+    public IConst getInitialValue(int offset) {
+        return initialValues.getOrDefault(offset,IValue.ZERO);
+    }
+
+    /**
+     * Defines the initial value at a certain field of this array.
+     * @param offset
+     * Non-negative number of fields before the target.
+     * @param value
+     * New value to be read at the start of each execution.
+     */
+    public void setInitialValue(int offset, IConst value) {
+        checkArgument(offset >= 0 && offset < size, "array index out of bounds");
+        initialValues.put(offset,value);
+    }
+
+    /**
+     * Updates the initial value at a certain field of this array.
+     * Resizes this array if the index exceeds the bounds.
+     * @param offset
+     * Non-negative number of fields before the target.
+     * @param value
+     * New value to be read at the start of each execution.
+     */
+    public void appendInitialValue(int offset, IConst value) {
+        checkArgument(offset >= 0, "array index out of bounds");
+        //The current implementation of Smack does not provide proper size information on static arrays.
+        //Instead, it indicates the size in the Boogie file by initializing each of the fields in a special procedure.
+        if(size <= offset) {
+            size = offset + 1;
+        }
+        initialValues.put(offset,value);
+    }
+
+    public IExpr add(int offset) {
+        checkArgument(0<=offset && offset<size, "array index out of bounds");
+        return offset == 0 ? this : new IExprBin(this,PLUS,new IValue(BigInteger.valueOf(offset),getPrecision()));
+    }
+
+    public Formula getLastMemValueExpr(SolverContext ctx, int offset) {
+        checkArgument(0<=offset && offset<size, "array index out of bounds");
+        String name = String.format("last_val_at_memory_%d_%d",index,offset);
+        return ctx.getFormulaManager().getIntegerFormulaManager().makeVariable(name);
     }
 
     @Override
-    public Formula getLastValueExpr(SolverContext ctx){
-        return toIntFormula(ctx);
+    public BigInteger getValue() {
+        return value != null ? value : BigInteger.valueOf(index);
     }
 
-    public Formula getLastMemValueExpr(SolverContext ctx){
-		String name = "last_val_at_memory_" + index;
-		return ctx.getFormulaManager().getIntegerFormulaManager().makeVariable(name);
+    @Override
+    public int getPrecision() {
```

Comment:
We have precision in `IExpr`, which `IConst` and `Address` inherit from.  In `development`, `Address` still calls the constructor of `IConst` with precision `-1`.  I think you removed dynamic precision from `Address`, which was implemented in a way that addresses to a boolean would be encoded as a boolean themselves.

Reply:
Ok, you are right ... I just removed it from the `ProgramBuilder`. But if `Address extends IConst` and `IConst` already has an implementation of `getPrecision()`, why do we need one in `Address`?

Reply:
That's because he changed `IConst`. It no longer has a precision variable. In fact, it is essentially an interface now.

Reply:
if we remove the `getPrecision()` method from `Address`, it would call the method of the parent (i.e. `IConst`) and still return `-1`.

If we treat `IConst` as an interface, we need to have more than one implementation, and right now since `Address extends IConst`, it doesn't seem this is the case.

Reply:
I introduced `IValue` because of this `extends` relationship, that surely is important in multiple parts of this project and difficult to remove.  I especially needed to change the return value of `Address#getValue()` in `Memory#fixAddressValues()`.  `Address#getPrecision()` now directly overrides `IExpr#getPrecision()`, which throws an `UnsupportedOperationException`.